### PR TITLE
[7X] Also set dependent array type encoding on ALTER TYPE ENCODING

### PR DESCRIPTION
**Pull request by kyeap-vmware**
Changing a type's default encoding using ALTER TYPE name SET DEFAULT
ENCODING should also set it's dependent array type to the same encoding.
This is the expected behavior if default compression encoding is
specified on CREATE TYPE. This also fixes pg_dump and pg_upgrade since
pg_dump relies on ALTER TYPE statements to set a type's default
encoding.

Without this change user defined types with their default encodings
specified on creation will set default encodings on the both type and
it's depedent array type. The ALTER TYPE SET DEFAULT ENCODING statement
from pg_dump would not restore the default encoding of the dependent
array type.

This change prohibits altering types without a dependent array type.
This was allowed under the following assumptions.
1. User defined types will always have dependent array types
2. It was fine that this prohibits some internal types that have no
   dependent array types from having their default encodings modified.
3. Forcing types and their dependent array to restore with the same
   default encoding will not break existing tables because the columns
   store their own encoding.

While this means some internal types now cannot have their default
encodings changed we will keep user defined types and their dependent
arrays types default encoding in sync.

pipeline: https://cm.ci.gpdb.pivotal.io/teams/main/pipelines/alter-encode

### CODE REVIEW

File: src/backend/catalog/pg_type.c [hunk 1]
```diff
@@ -158,15 +158,13 @@ remove_type_encoding(Oid typid)
...
-update_type_encoding(Oid typid, Datum typoptions)
+update_type_encoding(Relation pgtypeenc, Oid typid, Datum typoptions)
 {
-	Relation 	pgtypeenc;
 	ScanKeyData 	scankey;
 	SysScanDesc 	scan;
 	HeapTuple	tup;
 
 	/* SELECT * FROM pg_type_encoding WHERE typid = :1 FOR UPDATE */
-	pgtypeenc = heap_open(TypeEncodingRelationId, RowExclusiveLock);
```
Comment (jimmyyih) [inline src/backend/catalog/pg_type.c:169]: Actually, it seems that it's fine to leave the `heap_open` and `heap_close` here instead of moving it up to the caller.  The locking code will handle it okay.  We can revert the change to this function and then the patch will be significantly smaller.

File: src/backend/commands/typecmds.c [hunk 2]
```diff
@@ -3887,7 +3890,19 @@ AlterType(AlterTypeStmt *stmt)
 									 encoding, NULL, NULL,
 									 false,
 									 false);
-	update_type_encoding(typid, typoptions);
+
+	if (arrtypid == InvalidOid)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_OBJECT_DEFINITION),
+				 errmsg("Failed to get \"%s\" array type",
```
Comment (jimmyyih) [inline src/backend/commands/typecmds.c:3897]: The error message must start with a lowercase so s/Failed/failed/.  I would update the error message to something like this too: `failed to get the array type of type \"%s\"`.

File: src/backend/commands/typecmds.c [hunk 3]
```diff
@@ -3887,7 +3890,19 @@ AlterType(AlterTypeStmt *stmt)
 									 encoding, NULL, NULL,
 									 false,
 									 false);
-	update_type_encoding(typid, typoptions);
+
+	if (arrtypid == InvalidOid)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_OBJECT_DEFINITION),
```
Comment (jimmyyih) [inline src/backend/commands/typecmds.c:3896]: The error code should be ERRCODE_UNDEFINED_OBJECT for things that don't exist (e.g. found InvalidOid but expected a valid one).

File: src/test/regress/sql/AOCO_Compression.sql [hunk 4]
```diff
@@ -1779,3 +1779,14 @@ create table a_aoco_table_with_rle_type_and_invalid_compression_level(col int) W
 
 -- Check that callbacks are registered
 SELECT * FROM pg_compression WHERE compname='rle_type';
+
+-- Check ALTER TYPE SET DEFAULT ENCODING propagates to the dependent arrary type
```
Comment (jimmyyih) [inline src/test/regress/sql/AOCO_Compression.sql:1783]: s/arrary/array/

File: src/test/regress/sql/AOCO_Compression.sql [hunk 5]
```diff
@@ -1779,3 +1779,14 @@ create table a_aoco_table_with_rle_type_and_invalid_compression_level(col int) W
...
+DROP type if exists mood cascade;
+
+CREATE TYPE mood AS ENUM ('sad', 'ok', 'happy');
+ALTER TYPE public.mood SET DEFAULT ENCODING (compresstype=zlib, blocksize=65536, compresslevel=4);
```
Comment (jimmyyih) [inline src/test/regress/sql/AOCO_Compression.sql:1787]: Let's make the type name more unique so it's easier to search/debug if something fails (mood type is also created in hash_func test but is dropped at end of test).  Maybe call it `mood_encoded`.

**Review by jimmyyih — APPROVED** [review]
Nice commit message.  This looks good to me now.

**Comment by kalensk** [comment]
Maybe its just me, but it seems very useful to include this part of the commit message as a comment in the source code where the error occurs. This makes it easy for future developers to know "why" since in the future they may want to change that behavior and that message may help them understand some of the ramifications if they do.

```
This change prohibits altering types without a dependent array type.
This was allowed under the following assumptions.
1. User defined types will always have dependent array types
2. It was fine that this prohibits some internal types that have no
   dependent array types from having their default encodings modified.
3. Forcing types and their dependent array to restore with the same
   default encoding will not break existing tables because the columns
   store their own encoding.

While this means some internal types now cannot have their default
encodings changed we will keep user defined types and their dependent
arrays types default encoding in sync.
```
